### PR TITLE
feat: add IPAM options block for docker networks

### LINKED ISSUE #488: Support for passing IPAM options to ipam driver in docker_network
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Several IPAM drivers (`kuryr-libnetwork` for one) require passing options to the IPAM driver in addition to the regular driver. My particular use case requires that I be able to set the subnet pool and, so far as I can tell, it wants to be set for both the driver and IPAM driver. Being able to get at the IPAM driver options would benefit multiple other drivers that otherwise just work. 

I think a simple type map as is already used for `driver` is sufficient, and I think it would be best nested under `ipam_config`. Alternatively, there could be a separate `ipam_options` entry (as Compose does it).

### New or Affected Resource(s)

* docker_network

### Potential Terraform Configuration
Note that the references to `${openstack*}` items here are just filler and an example of my use case. 

```hcl
resource "docker_network" "kuryr_net" {
   attachable = true
   name       = "kuryr_net"
   driver     = "kuryr"
   options    = {
                    "neutron.pool.uuid"="${openstack_networking_subnetpool_v2.net_pool.id}"
                    "neutron.net.uuid"="${openstack_networking_network_v2.net_network.id}"
                }
   ipam_driver = "kuryr"
   ipam_config  {
                   gateway = "${openstack_networking_subnet_v2.net_subnet.gateway_ip}"
                   subnet  = "${openstack_networking_subnet_v2.net_subnet.cidr}"
                   options = { "neutron.net.uuid"="${openstack_networking_network_v2.net_network.id}" }
                }
}
```
or 
```hcl
resource "docker_network" "kuryr_net" {
   attachable = true
   name       = "kuryr_net"
   driver     = "kuryr"
   options    = {
                    "neutron.pool.uuid"="${openstack_networking_subnetpool_v2.net_pool.id}"
                    "neutron.net.uuid"="${openstack_networking_network_v2.net_network.id}"
                }
   ipam_driver = "kuryr"
   ipam_config  {
                   gateway = "${openstack_networking_subnet_v2.net_subnet.gateway_ip}"
                   subnet  = "${openstack_networking_subnet_v2.net_subnet.cidr}"
                }
   ipam_options = { "neutron.net.uuid"="${openstack_networking_network_v2.net_network.id}" }
}
```

### References
* [Kuryr-libnetwork documentation](https://docs.openstack.org/kuryr-libnetwork/latest/readme.html#kuryr-libnetwork)
* [Compose example of options](https://docs.docker.com/compose/compose-file/#ipam)

### PR DESCRIPTION
**Pull request by lenox-joseph**
Extends the docker provider to be able to also set IPAM options for drivers.  This is highly useful for more exotic network drivers (such as kuryr-libnetwork).
Fixes #488

Here's a snippet of the created output. All I did was add to the schema and plug in the IPAM structure that the API exposes.
```
xxxx@devstack-host-2:~/pingcont$ docker network inspect ea7c9f982e74
[
    {
        "Name": "net_kuryr_1",
        "Id": "ea7c9f982e74a1f7813142e6acec02b829f68102102555a037a6327d9bc73d20",
        "Created": "2022-12-16T23:37:45.885531072Z",
        "Scope": "local",
        "Driver": "kuryr",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "kuryr",
            "Options": {
                "neutron.net.uuid": "267aaee5-27a8-404c-998d-223c1437240c",
                "neutron.pool.uuid": "ee346cfa-7082-46b6-9b92-ba03763252c7"
            },
            "Config": [
```

### CODE REVIEW

**Review by Junkern — COMMENTED** [review]
Nice and quick work! Whenever we add a feature to the `*Create` function (in your case the `resourceDockerNetworkCreate`) we also have to add that feature to the `*Read` function.

Basically the same as with the existing `ipam_config` in https://github.com/kreuzwerker/terraform-provider-docker/pull/491/files#diff-ae39f63508e617ca4403475ba2466d710377123ece15be0a8224793de5d5bc6bL181 ,  otherwise there will always be a diff when running `terraform plan` the next time

**Comment by lenox-joseph** [comment]
> Nice and quick work! Whenever we add a feature to the `*Create` function (in your case the `resourceDockerNetworkCreate`) we also have to add that feature to the `*Read` function.
> 
> Basically the same as with the existing `ipam_config` in https://github.com/kreuzwerker/terraform-provider-docker/pull/491/files#diff-ae39f63508e617ca4403475ba2466d710377123ece15be0a8224793de5d5bc6bL181 , otherwise there will always be a diff when running `terraform plan` the next time

Done.

**Comment by Junkern** [comment]
Awesome! And if you have 2 more minutes to [update-the-documentation](https://github.com/kreuzwerker/terraform-provider-docker/blob/master/CONTRIBUTING.md#update-the-documentation) that would be great :)

**Comment by lenox-joseph** [comment]
> Awesome! And if you have 2 more minutes to [update-the-documentation](https://github.com/kreuzwerker/terraform-provider-docker/blob/master/CONTRIBUTING.md#update-the-documentation) that would be great :)

Should be done, although I'm getting complaints about the commit titles when trying to rebase and clean up commit messages.

**Comment by lenox-joseph** [comment]
followup @Junkern the dev system I'm working on is behind a proxy and even specifying the proxy via `-e` on the `docker run` command is giving weird results for links that didn't change. As far as I can tell everything should be fine.

**Review by Junkern — APPROVED** [review]

File: internal/provider/resource_docker_network.go [hunk 1]
```diff
@@ -96,6 +96,13 @@ func resourceDockerNetwork() *schema.Resource {
 				Optional:    true,
 				ForceNew:    true,
 			},
+			"ipam_options": {
+				Type:        schema.TypeMap,
+				Description: "Only available with bridge networks. See [bridge options docs](https://docs.docker.com/engine/reference/commandline/network_create/#bridge-driver-options) for more details.",
+				Optional:    true,
+				ForceNew:    true,
+				Computed:    true,
```
Comment (Junkern) [inline internal/provider/resource_docker_network.go:104]: Ah, this should be `false` (or just remove it completely)

**Review by Junkern — CHANGES_REQUESTED** [review]
Really small change

**Comment by Junkern** [comment]
And no need to rebase, I always to squash merges :)

File: internal/provider/resource_docker_network.go (hunk 1 above)
Comment (lenox-joseph) [reply internal/provider/resource_docker_network.go:104]: `ForceNew` and `Computed`? The snippet's not very clear. I am going to revise the text though to match what I put in the docs.

Comment (Junkern) [reply internal/provider/resource_docker_network.go:104]: The docs are auto-generated, so you only have to modify the text here.
And simply remove the `Computed` property
```suggestion
```

**Comment by Junkern** [comment]
I have no idea why the docs checks errors. Will merge for now and fix that right before the next release :)

**Review by Junkern — APPROVED** [review]

**Pull request merged**